src/backend/distributed/planner/multi_router_planner.c

```diff
@@ -807,253 +844,383 @@ RouterSelectTask(Query *query)
-	 * Queries which have subqueries, or tablesamples in FROM clauses are rejected later
-	 * during RangeTblEntry checks.
-	 */
-	if (query->hasSubLinks == true || query->cteList != NIL || query->hasForUpdate ||
-		query->hasRecursive)
-	{
-		return false;
-	}
+		/* retrieve all active shard placements for this shard */
+		shardPlacementList = FinalizedShardPlacementList(shardId);
 
-#if (PG_VERSION_NUM >= 90500)
-	if (query->groupingSets)
-	{
-		return false;
-	}
-#endif
+		/* create nodeName:nodePort tuples for each placement */
+		foreach(shardPlacementCell, shardPlacementList)
+		{
+			ShardPlacement *shardPlacement = (ShardPlacement *) lfirst(
+				shardPlacementCell);
+			StringInfo nodeInfo = makeStringInfo();
+			appendStringInfo(nodeInfo, "%s:%u", shardPlacement->nodeName,
+							 shardPlacement->nodePort);
+			shardNodeList = lappend(shardNodeList, nodeInfo);
+		}
 
-	/* only hash partitioned tables are supported */
-	distributedTableId = ExtractFirstDistributedTableId(query);
-	partitionColumn = PartitionColumn(distributedTableId, rangeTableId);
-	partitionMethod = PartitionMethod(distributedTableId);
+		/*
+		 * Perform placement pruning based on string matching on nodeName:nodePort
+		 * tuples. We start pruning from all placements of the first relation's shard.
+		 * Then for each relation's shard, we compute intersection of the new shards
+		 * placement with existing placement list. This operation could have been
+		 * done using other methods, but since we do not expect very high replication
+		 * factor, iterating over a list and making string comparisons should be the
+		 * fastest way around.
+		 */
+		if (firstShard)
+		{
+			firstShard = false;
+			workerNodeList = shardNodeList;
+			placementList = shardPlacementList;
+		}
+		else
+		{
+			List *existingWorkerNodeList = workerNodeList;
+			ListCell *existingWorkerNodeCell = NULL;
+			ListCell *shardNodeCell = NULL;
 
-	if (partitionMethod != DISTRIBUTE_BY_HASH)
-	{
-		return false;
-	}
+			/* reset active placement list and active worker node list */
+			placementList = NIL;
+			workerNodeList = NIL;
 
-	/* extract range table entries */
-	ExtractRangeTableEntryWalker((Node *) query, &rangeTableList);
+			forboth(shardNodeCell, shardNodeList, shardPlacementCell, shardPlacementList)
+			{
+				ShardPlacement *shardPlacement =
+					(ShardPlacement *) lfirst(shardPlacementCell);
+
+				foreach(existingWorkerNodeCell, existingWorkerNodeList)
+				{
+					StringInfo existingNodeInfo =
+						(StringInfo) lfirst(existingWorkerNodeCell);
+					StringInfo newNodeInfo = (StringInfo) lfirst(shardNodeCell);
+
+					/*
+					 * Append placement to placement list if it is already in existing
+					 * placement list.
+					 */
+					if (strncmp(existingNodeInfo->data, newNodeInfo->data,
+								existingNodeInfo->len) == 0)
+					{
+						placementList = lappend(placementList, shardPlacement);
+						workerNodeList = lappend(workerNodeList, existingNodeInfo);
+					}
+				}
+			}
```

Comment:
Some of the queries motivating this have dozens to low hundreds table references, so we're not talking just peanuts. And it's not only the string comparison, but also the the significant of memory allocations you're performing for each combination.